contracts/exchangeIssuance/ExchangeIssuanceLeveraged.sol

```diff
@@ -0,0 +1,1159 @@
+    * 
+    * @param _weth                  Address of wrapped native token
+    * @param _quickRouter           Address of quickswap router
+    * @param _sushiRouter           Address of sushiswap router
+    * @param _setController         SetToken controller used to verify a given token is a set
+    * @param _debtIssuanceModule    DebtIssuanceModule used to issue and redeem tokens
+    * @param _addressProvider       Address of DebtIssuanceModule used to issue and redeem tokens
+    */
+    constructor(
+        address _weth,
+        IUniswapV2Router02 _quickRouter,
+        IUniswapV2Router02 _sushiRouter,
+        ISwapRouter _uniV3Router,
+        IController _setController,
+        IDebtIssuanceModule _debtIssuanceModule,
+        address _addressProvider
+    )
+        public
+        FlashLoanReceiverBaseV2(_addressProvider)
+        DEXAdapter(_weth, _quickRouter, _sushiRouter, _uniV3Router)
+    {
+        setController = _setController;
+        debtIssuanceModule = _debtIssuanceModule;
+
+        WETH = _weth;
+    }
+
+    /* ============ External Functions ============ */
+
+
+    /**
+    * Returns the collateral / debt token addresses and amounts for a leveraged index 
+    *
+    * @param _setToken              Address of the SetToken to be issued / redeemed
+    * @param _setAmount             Amount of SetTokens to issue / redeem
+    * @param _isIssuance            Boolean indicating if the SetToken is to be issued or redeemed
+    *
+    * @return Struct containing the collateral / debt token addresses and amounts
+    */
+    function getLeveragedTokenData(
+        ISetToken _setToken,
+        uint256 _setAmount,
+        bool _isIssuance
+    )
+        isSetToken(_setToken)
+        external 
+        view
+        returns (LeveragedTokenData memory)
+    {
+        return _getLeveragedTokenData(_setToken, _setAmount, _isIssuance);
+    }
+
+    /**
+     * Runs all the necessary approval functions required for a given ERC20 token.
+     * This function can be called when a new token is added to a SetToken during a
+     * rebalance.
+     *
+     * @param _token    Address of the token which needs approval
+     */
+    function approveToken(IERC20 _token) external {
+        _approveToken(_token);
+    }
+
+
+    /**
+     * Trigger redemption of set token to pay the user with Eth
+     *
+     * @param _setToken               Set token to redeem
+     * @param _setAmount              Amount to redeem
+     * @param _minAmountOutputToken   Minimum amount of ETH to send to the user
+     * @param _exchange               Exchange to use in swap from debt to collateral token
+     */
+    function redeemExactSetForETH(
+        ISetToken _setToken,
+        uint256 _setAmount,
+        uint256 _minAmountOutputToken,
+        Exchange _exchange,
+        SwapData memory _swapDataCollateralForDebt,
+        SwapData memory _swapDataOutputToken
+    )
+        isSetToken(_setToken)
+        public
```

Comment:
Is there any reason these aren't external? We usually make everything (except the constructor) external by convention. On older versions of solidity its actually more gas efficient, but now its just sort of the standard. Also public/external have the same inherited visibility thing as private/internal which I explained earlier.

Reply:
Hmm that's interesting... I'll have to look into that one

Reply:
Seems my memory was failing me. Using `memory` structs with `external` access modifier works fine. Adjusted accordingly

Reply:
Calldata structs do result in `UnimplementedError` though, as I just verified. So I'll keep the `swapData` structs as `memory` parameters.